HW4 Integration

jinglinstella opened the pull request:
# Pull Request Description

## Overview
[Provide a brief overview of what this PR accomplishes]

## Type of Change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvements
- [ ] Test updates

## Changes Made
[Describe the changes you've made in detail]

### Added
- [List new features/functionality added]

### Changed
- [List modifications to existing functionality]

### Removed
- [List removed features/functionality]

## Testing
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [ ] All tests passing locally
- [ ] Coverage requirements met (90%+)

## Additional Information
- [ ] mypy static type checking passes
- [ ] ruff formatting and linting passes
- [ ] Documentation has been updated
- [ ] Component specification updated (if needed)

## Screenshots (if applicable)
[Add screenshots here if UI changes were made]

## Related Issues
- Closes #[issue number]
- References #[issue number]

## Deployment Notes
[Any special considerations for deploying this change]

## Reviewer Checklist
- [ ] Code follows project style guidelines
- [ ] Code is well-documented
- [ ] Changes are covered by tests
- [ ] Type hints are properly used
- [ ] Error handling is appropriate
- [ ] Performance implications considered
- [ ] Security implications considered

## Self Review Checklist
Before requesting review, I have:
- [ ] Performed self-review of my code
- [ ] Commented on hard-to-understand areas
- [ ] Updated relevant documentation
- [ ] Verified all tests pass
- [ ] Verified code formatting and linting pass
- [ ] Verified type checking passes

zz2548 committed:
Changed ai_bot_channel.py

Changed ai_bot_channel.py to use logging
Configured flask app to have logging

zz2548 committed:
Use singleton for issue tracker

jinglinstella committed:
fix cicid

jinglinstella committed:
fix cicid

jinglinstella committed:
fix cicid

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update yml

jinglinstella committed:
update toml

jinglinstella committed:
fix no module found error

jinglinstella committed:
fix loadChannels

jinglinstella committed:
Update issue_tracker submodule pointer after fix

jinglinstella committed:
update frontend

jinglinstella committed:
update db before merging

jinglinstella committed:
Merge branch 'hw4-jt3' into hw4-jt

Cybercricetus committed:
Update README.md

jinglinstella committed:
add integration test

jinglinstella committed:
update toml

jinglinstella committed:
update yml

jinglinstella committed:
update yml

jinglinstella committed:
update yml

jinglinstella committed:
update path

jinglinstella committed:
update gitmodules

jinglinstella committed:
update yml

jinglinstella committed:
Update submodule issue_tracker to latest commit for typing fixes

jinglinstella committed:
Update issue_tracker submodule pointer after fixing URL

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update toml

jinglinstella committed:
update readme

Mirap9615 commented:
# PR Feedback: Team 14 
**Strengths**
- Clear Project Structure, with clear intention of modularity in mind, especially within /src
- Good implementation of modern tooling, testing, CI/CD workflows 
- Detailed and well structured readme, overview, setup steps, feature summary, and comprehensive API documentation all included
- API covers essential chat features, direct messaging and AI bot integration

**Areas for Improvement**
- insufficient implementation of session management / token-based auth (/login, etc), too much dependence on unverified user_ids
- hardcoded config vars (database URI, track modifications)
- incomplete / incorrect features (join_channel, list_channel_users - should include users that have joined but haven't yet sent a message)
- more robust error handling, especially for runtime errors, such as for issues with database communication, invalid payloads. Could further implement flask's error handlers 
- add more thorough type annotations / docstrings for functions

**Further Enhancements**
- db initialization commands could be more graceful with flask cli
- could improve testing breadth by including raw HTTP requests to flask app in addition to requests via API
- lazy import with AiBotChannel, clarify if for lazy or else move to top of file 
- commented out pre commit test hooks as well as leftover configuration from earlier homeworks such as reference to src/calculator/test
- could make ai bot trigger (channel.name == "ai-helpdesk") more dynamic / configurable 
- potential issue with dm channel naming, sort user ids to ensure uniqueness

unknownblueguy6 commented:
I think you have received great feedback from Team 14. They have addressed important improvements as well as enhancements you should implement, especially regarding hardcoded config variables and docstrings.

zz2548 committed:
Remove lazy loading

zz2548 committed:
Moving import up

ml378 committed:
Added docstrings to ai_bot_channel.py

zz2548 committed:
Using dotenv for environment variable

Using dotenv for environment variable, also removed one instance of lazy import
Added server config settings in dotenv
AI Bot features now configurable with dotenv

zz2548 committed:
Merge branch 'hw4-jt-int-test' of https://github.com/zz2548/Team-9-Technology-Template-Repository into hw4-jt-int-test

ml378 committed:
Added docstrings to _impl.py

zz2548 committed:
Added flask CLI commands

In powershell:
$env:FLASK_APP = "app.py"
flask init-db
flask seed-ai-bot
flask drop-db

zz2548 committed:
Merge branch 'hw4-jt-int-test' of https://github.com/zz2548/Team-9-Technology-Template-Repository into hw4-jt-int-test

zz2548 committed:
Use lowercase name for db_type and db_path

ml378 committed:
Updated .pre-commit-config.yaml

zz2548 committed:
Updated readme

zz2548 committed:
Merge branch 'hw4-jt-int-test' of https://github.com/zz2548/Team-9-Technology-Template-Repository into hw4-jt-int-test

ml378 committed:
Added docstrings to test_channel.py

zz2548 committed:
Added docstring and type annotation for app.py

ml378 committed:
Added docstrings to test_chat_client.py

zz2548 committed:
Added new relationship between user and channel

Creating a proper join table (ChannelUserModel) to track the relationship between users and channels
Updating the join_channel endpoint to create records in this join table
Modifying the list_channel_users endpoint to query the join table instead of inferring membership from messages

ml378 committed:
Added a top-level docstring to test_message_flow.py

ml378 committed:
Addded docstrings to test_userflow.py

zz2548 committed:
Allow proper fetching of user name

zz2548 committed:
Merge branch 'hw4-jt-int-test' of https://github.com/zz2548/Team-9-Technology-Template-Repository into hw4-jt-int-test

ml378 committed:
Added null checks for JSON extraction to app.py

ml378 committed:
Added flask error handlers to app.py

zz2548 committed:
Update dependencies in pyproject.toml

Moved the actual application dependencies from the dev section to the main dependencies:

flask
flask-sqlalchemy
flask-cors
openai
python-dotenv
aiohttp
httpx

to install: use uv sync --group dev

zz2548 committed:
Merge branch 'hw4-jt-int-test' of https://github.com/zz2548/Team-9-Technology-Template-Repository into hw4-jt-int-test

ml378 committed:
Added inclusion of raw HTTP requests to app.py

ml378 committed:
Update app.py

zz2548 committed:
Added authentication, updated integration test

zz2548 committed:
Update test_user.py

zz2548 committed:
Update auth

zz2548 committed:
Updated docstrings and return annotation

zz2548 committed:
update type annotation

zz2548 committed:
Update linting rules

zz2548 committed:
Update ci config

zz2548 committed:
Revert "Update ci config"

This reverts commit 9ae3217d862be02362dd17342512afb88e6367b1.

zz2548 committed:
Revert ci changes

zz2548 committed:
Update ci config to init db

zz2548 committed:
update circleci config

zz2548 committed:
update ci config

zz2548 committed:
update ci config

zz2548 committed:
Merge branch 'hw4-auth' into hw4-jt-int-test

ml378 committed:
Update README.md
